### PR TITLE
Add vscode Ruby LSP config and documentation

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR updates the VS Code configuration to use the Ruby LSP extension instead of the old one which has been deprecated. I've also added some extra documentation to help explain how to use it. This setup seems to work for me, but I would love some other VS Code users to check it and make sure it works for them too.

The basis of the approach I've taken is documented in the extension docs [here](https://github.com/Shopify/ruby-lsp/blob/main/vscode/README.md?tab=readme-ov-file#using-a-custom-gemfile).

CC: @p-linnane @krehel @Bo98

### CODE REVIEW

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+Library/Homebrew/.ruby-version
```
Comment (carlocab) [inline .ruby-version:1]: IIRC this was where we put the `.ruby-version` file initially, but this breaks `rbenv` for scripts that use `#!/usr/bin/env ruby`.

Comment (carlocab) [reply .ruby-version:1]: See #17390.

Comment (MikeMcQuaid) [reply .ruby-version:1]: Could consider adding this to `.gitignore` and having some Homebrew script to automatically make this symlink.

File: .vscode/ruby-lsp/Gemfile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+source 'https://rubygems.org'
+
+gem 'ruby-lsp'
+eval_gemfile(File.expand_path('../../Library/Homebrew/Gemfile', __dir__))
```
Comment (MikeMcQuaid) [inline .vscode/ruby-lsp/Gemfile:4]:
```suggestion
source "https://rubygems.org"

gem "ruby-lsp"
eval_gemfile(File.expand_path("../../Library/Homebrew/Gemfile", __dir__))
```

File: .vscode/ruby-lsp/Gemfile.lock [hunk 3]
```diff
@@ -0,0 +1,206 @@
+GEM
```
Comment (MikeMcQuaid) [inline .vscode/ruby-lsp/Gemfile.lock:1]: Might be nice to `.gitignore` this file so it doesn't need kept up-to-date by Dependabot etc.

File: docs/VSCode.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+- [`rbenv`](https://formulae.brew.sh/formula/rbenv)
+  - `brew install --formula rbenv`
+
+## Setup `rbenv`
+
+Once `rbenv` is set up correctly (see the [installation instructions](https://github.com/rbenv/rbenv#installation)), install the correct Ruby version:
+
+```bash
+rbenv install "$(cat "$(brew --repository)/.ruby-version")"
+```
```
Comment (MikeMcQuaid) [inline docs/VSCode.md:20]: Would recommend doing a `brew install ruby@X` and `brew rbenv-sync` here instead.

**Review by MikeMcQuaid — COMMENTED** [review]
Great work @Rylan12! Would love to see this working nicely.

File: .vscode/ruby-lsp/Gemfile (hunk 2 above)
Comment (MikeMcQuaid) [reply .vscode/ruby-lsp/Gemfile:4]: Also think this could be added to our main `Gemfile` in a specific `vscode` group?

File: .ruby-version (hunk 1 above)
Comment (Bo98) [reply .ruby-version:1]: Consider investigating setting `rubyLsp.customRubyCommand` (requires setting `rubyLsp.rubyVersionManager` to `custom`) to Portable Ruby or a shim that uses ruby.sh. That will likely avoid the need for this.

Multi-root workspaces may also work here, though not sure if that's what we want.

**Comment by Rylan12** [comment]
Okay I pushed a better solution which originally I couldn't get to work right, but I found a way to do it.

Now we can use portable ruby as the ruby for the LSP which avoids needing a second `Gemfile` at all. The only step necessary to use this now is to run `brew install-bundler-gems --add-groups style,vscode`.

I've removed the docs for it because it seems silly to have a whole page for that one line, but maybe it can go somewhere else in the docs. Once I've got the sorbet LSP set up, maybe there will be enough info for a joint page.

**Comment by MikeMcQuaid** [comment]
> I've removed the docs for it because it seems silly to have a whole page for that one line, but maybe it can go somewhere else in the docs.

It might be nice to figure out if you can convince VSCode to run that command automatically for you on launch. I know you can get it to run "tasks" when you open a workspace. Let me know if you need more syntax pointers there, I've done it on another project.

File: .vscode/settings.json [hunk 5]
```diff
@@ -6,8 +6,17 @@
   "files.trimTrailingWhitespace": true,
   "files.insertFinalNewline": true,
   "files.trimFinalNewlines": true,
-  "ruby.useLanguageServer": true,
-  "ruby.rubocop.executePath": "Library/Homebrew/shims/gems/",
+  "rubyLsp.rubyVersionManager": {
+    "identifier": "custom"
+  },
+  "rubyLsp.customRubyCommand": "PATH=${workspaceFolder}/Library/Homebrew/vendor/portable-ruby/current/bin:$PATH",
```
Comment (MikeMcQuaid) [inline .vscode/settings.json:12]: I wonder if we could even have a shim for the custom Ruby command that automatically installs the gems for you?

**Review by MikeMcQuaid — APPROVED** [review]
Looks good as-is, good to merge as-is and iterate! Thanks @Rylan12!

**Comment by Bo98** [comment]
You may find this snippet from an old `git stash` useful:

```
+  "rubyLsp.customRubyCommand": "source \"../../.vscode/ruby-lsp/activate.sh\"",
```

```sh
homebrew_prefix="$(cd "$(dirname "$0")"/../../ && pwd)/"

"${homebrew_prefix}/bin/brew" install-bundler-gems --add-groups=style,vscode

export PATH="${homebrew_prefix}/Library/Homebrew/vendor/portable-ruby/current/bin:${PATH}"
export BUNDLE_WITH="style vscode"
```

File: .vscode/settings.json [hunk 6]
```diff
@@ -6,8 +6,17 @@
...
+  "rubyLsp.bundleGemfile": "Library/Homebrew/Gemfile",
+  "rubyLsp.formatter": "rubocop",
+  "[ruby]": {
+    "editor.defaultFormatter": "Shopify.ruby-lsp",
+    "editor.formatOnSave": true,
```
Comment (Bo98) [inline .vscode/settings.json:17]: `editor.formatOnPaste` might be useful too. Or at least I had that locally.

**Comment by Bo98** [comment]
> Once I've got the sorbet LSP set up, maybe there will be enough info for a joint page.

Not for this PR but here's also some settings I got on this before, given I managed to get it working better than ruby-lsp:

```diff
+  "sorbet.enabled": true,
+  "sorbet.lspConfigs": [
+    {
+      "id": "default",
+      "name": "Brew Typecheck",
+      "description": "Default configuration",
+      "cwd": "${workspaceFolder}",
+      "command": [
+        "./bin/brew",
+        "typecheck",
+        "--lsp"
+      ]
+    }
+  ],
```

```diff
diff --git a/Library/Homebrew/dev-cmd/typecheck.rb b/Library/Homebrew/dev-cmd/ty
pecheck.rb
index a105d4ac71..00554f7dfc 100644
--- a/Library/Homebrew/dev-cmd/typecheck.rb
+++ b/Library/Homebrew/dev-cmd/typecheck.rb
@@ -21,6 +21,8 @@ module Homebrew
       switch "--suggest-typed",
              depends_on:  "--update",
              description: "Try upgrading `typed` sigils."
+      switch "--lsp",
+             description: "Start the Sorbet LSP server."
       flag   "--dir=",
              description: "Typecheck all files in a specific directory."
       flag   "--file=",
@@ -30,6 +32,9 @@ module Homebrew
                           "in their paths (relative to the input path passed to Sorbet)."
 
       conflicts "--dir", "--file"
+      conflicts "--lsp", "--update"
+      conflicts "--lsp", "--update-all"
+      conflicts "--lsp", "--fix"
 
       named_args :none
     end
@@ -91,6 +96,11 @@ module Homebrew
         srb_exec << "--autocorrect"
       end
 
+      if args.lsp?
+        srb_exec << "--lsp"
+        srb_exec << "--disable-watchman" unless which("watchman", ORIGINAL_PATHS)
+      end
+
       srb_exec += ["--ignore", args.ignore] if args.ignore.present?
       if args.file.present? || args.dir.present?
         cd("sorbet") do
```

**Comment by Rylan12** [comment]
Thanks for all the tips. I'm going to merge as-is for now to get this in, and will look into the shim/task path in the meantime.

I was able to get sorbet working with your suggestion, thanks @Bo98! I'll open a PR for that, too.

**Pull request merged**